miner/src/service_transaction_checker.rs

```diff
@@ -16,37 +16,111 @@
 
 //! A service transactions contract checker.
 
-use call_contract::{CallContract, RegistryInfo};
+use std::collections::BTreeMap;
+use call_contract::{RegistryInfo, CallContract};
+use chain_info::ChainInfo;
 use types::ids::BlockId;
 use types::transaction::SignedTransaction;
 use ethabi::FunctionOutputDecoder;
+use ethereum_types::Address;
+use parking_lot::RwLock;
 
 use_contract!(service_transaction, "res/contracts/service_transaction.json");
 
 const SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME: &'static str = "service_transaction_checker";
 
+/// Certified addresses cache
+#[derive(Default)]
+struct CertifiedAddressesCache {
+	cache: RwLock<BTreeMap<Address, bool>>,
+	last_block: u64
+}
+
 /// Service transactions checker.
 #[derive(Default, Clone)]
-pub struct ServiceTransactionChecker;
+pub struct ServiceTransactionChecker {
+	certified_addresses_cache: CertifiedAddressesCache
```

Comment:
This means that cache is present only for a single instance. We create a new instance every time using `ServiceTransactionChecker::default()` both in `transact_contract` and `import_block` so in the first case the cache is always dropped after the check and in the second case we `refresh_cache` that is empty.

To make it work we either have to make the cache static (not really recommended) or hold a single instance of cache for the whole time. (either by sharing one instance of `ServiceTransactionChecker` or by passing the cache as `Arc<RwLock<..>>` in the `ServiceTransactionChecker::new()` function (so we get rid of `Default`).

To be honest I think I'd prefer to have a single instance (or rather `Arc<ServiceTxChecker>`; ideally shared between miner and client).

Reply:
and btw `HashMap` should be fine instead of `BTreeMap`.